prinz/src/main/scala/pl/touk/nussknacker/prinz/util/collection/immutable/MapMultiset.scala

```diff
@@ -0,0 +1,132 @@
+package pl.touk.nussknacker.prinz.util.collection.immutable
+
+import scala.collection._
+
+object MapMultiset {
+
+  def empty[T](): MapMultiset[T] =
+    new MapMultiset[T](Map.empty[T, Int].withDefaultValue(0))
+
+  def apply[T](elements: T*): MapMultiset[T] = {
+    val builder = new MapMultiSetBuilder[T]()
+    builder ++= elements
+    builder.result()
+  }
+
+  def apply[T](elements: Iterable[T]): MapMultiset[T] = {
+    val builder = new MapMultiSetBuilder[T]()
+    builder ++= elements
+    builder.result()
+  }
+
+  def apply[T](delegate: Map[T, Int]): MapMultiset[T] =
+    new MapMultiset[T](delegate.filter{ case (_, count) => count > 0 })
+
+  class MapMultiSetBuilder[T]() extends mutable.Builder[T, MapMultiset[T]] {
+    private val elements = new mutable.HashMap[T, Int].withDefaultValue(0)
+
+    override def +=(element: T): MapMultiSetBuilder.this.type = {
+      val count = elements(element)
+      elements.update(element, count + 1)
+      this
+    }
+
+    override def clear(): Unit = elements.clear()
+
+    override def result(): MapMultiset[T] = {
+      new MapMultiset[T](Map.empty[T, Int].withDefaultValue(0) ++ elements)
+    }
+  }
+
+}
+
+class MapMultiset[T](private val delegate: Map[T, Int]) {
```

Comment:
The main purpose of this file is this class so let's make it toplevel

Reply:
you also use Hash version of delegate so it should be reperesnted somehow in name (to tell user not to use it with no has defined)